Can’t use c and c++ mgems together

Hi there,

ive switched one of my widely used mgems to c++, now i cant mix that with .c extensions anymore because throw.h then gets compiled with gcc but sees these command line arguments:
```gcc
In file included from mruby-c-ares/mruby/src/error.c:17:
mruby-c-ares/mruby/include/mruby/throw.h:26:3: error: #error "need to be compiled with C++ compiler"
   26 | # error "need to be compiled with C++ compiler"
      |   ^~~~~
rake aborted!
Command failed with status (1): [gcc -E -P -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -g3 -O0 -fno-omit-frame-pointer -g -ggdb -Og -g3 -O0 -fexceptions -DMRB_DEBUG -DMRB_DEBUG -DMRB_USE_CXX_EXCEPTION -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DMRB_USE_SET -I"mruby-c-ares/mruby/include" -I"mruby-c-ares/mruby/build/host/include" -DMRB_PRESYM_SCANNING -o "mruby-c-ares/mruby/build/host/src/error.pi" "mruby-c-ares/mruby/src/error.c"]
```
when i add that c++ mgem to a build_config.rb file and add that with conf.gem mgem: it works just fine.

reproduction

create a mruby c extension, add empty gem_final and gem_init functions, add spec.add_dependency 'mruby-c-ext-helpers' to its mrbgem.rake, compile it.

this is driving me nuts, i have just one .cpp file anywhere in my chain and this happens.

```bash
In file included from /home/hendrik/code/totally-normal-keyboard/mruby/src/error.c:17:
/home/hendrik/code/totally-normal-keyboard/mruby/include/mruby/throw.h:26:3: error: "need to be compiled with C++ compiler"
   26 | # error "need to be compiled with C++ compiler"
      |   ^
1 error generated.
rake aborted!
Command failed with status (1): [clang -E -P -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -Wzero-length-array -g3 -O0 -Og -g -fno-omit-frame-pointer -march=armv8-a -fexceptions -DMRB_DEBUG -DMRB_USE_CXX_EXCEPTION -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DMRB_USE_SET -I"/home/hendrik/code/totally-normal-keyboard/mruby/include" -I"/home/hendrik/code/totally-normal-keyboard/mruby/build/debug/include" -DMRB_PRESYM_SCANNING -o "/home/hendrik/code/totally-normal-keyboard/mruby/build/debug/src/error.pi" "/home/hendrik/code/totally-normal-keyboard/mruby/src/error.c"]
```

here is my build config
```ruby
MRuby::Build.new('debug') do |conf|
  conf.toolchain :clang
  conf.enable_debug
  conf.gembox 'full-core'

  conf.cc.flags << '-Og' << '-g' << '-fno-omit-frame-pointer' << '-march=armv8-a'
  conf.cxx.flags << '-Og' << '-g' << '-std=c++20' << '-fno-omit-frame-pointer' << '-march=armv8-a'

  conf.gem File.expand_path(File.dirname(__FILE__))
end

MRuby::Build.new('release') do |conf|
  conf.toolchain :clang
  conf.gembox 'full-core'

  conf.cc.flags << '-Os' << '-flto' << '-ffunction-sections' << '-fdata-sections' << '-march=armv8-a'
  conf.cxx.flags << '-Os' << '-std=c++20' << '-flto' << '-ffunction-sections' << '-fdata-sections' << '-march=armv8-a'
  conf.linker.flags << '-Wl,--gc-sections' << '-Wl,--threads=4' << '-flto' << '-fuse-ld=lld'

  conf.gem File.expand_path(File.dirname(__FILE__))
end
```
and here the mrbgem.rake
```rake
MRuby::Gem::Specification.new('totally-normal-keyboard') do |spec|
  spec.license = 'AGPL-3.0'
  spec.authors = 'Hendrik Beskow'
  spec.add_dependency 'mruby-io-uring'

  spec.bins = %w(tnk)
end
```

you just need to #include <mruby.h> in any cpp file and this happens.

do you need to guard that around extern "C" { #include <mruby.h> } ?

--- a/Rakefile
+++ b/Rakefile
@@ -17,6 +17,11 @@ require "mruby/build"
 MRUBY_CONFIG = MRuby::Build.mruby_config_path
 load MRUBY_CONFIG
 
+# set up all gems
+MRuby.each_target do
+  gems.setup(self) if enable_gems?
+end
+
 # load basic rules
 MRuby.each_target do |build|
   build.define_rules

--- a/doc/guides/mrbgems.md
+++ b/doc/guides/mrbgems.md
@@ -310,6 +310,28 @@ For example: when B depends on C and A depends on B, A will get include paths ex
 Exported `include_paths` are automatically appended to GEM local `include_paths` by rake.
 You can use `spec.export_include_paths` accessor if you want more complex build.
 
+### Settings for GEM build commands/tasks
+
+When the block argument passed to `MRuby::Gem::Specification.new` is executed,
+the GEM build commands/tasks for the `MRuby::Build` instance may not yet be finalized.
+In most cases, modifying the GEM build commands/tasks within the block passed to
+`MRuby::Gem::Specification.new` is not a problem.
+
+However, you may need to perform GEM build commands/tasks after the GEM build
+commands/tasks for the `MRuby::Build` instance have been finalized.
+In such cases, you can achieve this by passing a block argument to
+`MRuby::Gem::Specification#build_settings` within the block passed to
+`MRuby::Gem::Specification.new`.
+
+```ruby
+spec.build_settings do
+  spec.cc.flags << "-any_flags"
+end
+```
+
+**NOTE**: Using the `build_settings` method will cause GEM's all build command settings
+directly written in the block passed to `MRuby::Gem::Specification.new` to be ignored.
+
 ## C Extension
 
 mruby can be extended with C. This is possible by using the C API to

--- a/lib/mruby/gem.rb
+++ b/lib/mruby/gem.rb
@@ -53,10 +53,8 @@ module MRuby
         return if defined?(@bins)  # return if already set up
 
         MRuby::Gem.current = self
-        MRuby::Build::COMMANDS.each do |command|
-          instance_variable_set("@#{command}", @build.send(command).clone)
-        end
-        @linker.run_attrs.each(&:clear)
+        reset_commands # for backward compatibility, reset the commands from the beginning.
+        @build_settings = nil
 
         @rbfiles = Dir.glob("#{@dir}/mrblib/**/*.rb").sort
         @objs = srcs_to_objs("src")
@@ -192,6 +190,19 @@ module MRuby
         end
       end
 
+      def build_settings(&blk)
+        @build_settings = blk
+      end
+
+      def setup_build
+        if @build_settings
+          # by this point, build.cc or other commands may have been modified.
+          # therefore, reset the commands again before calling build_settings.
+          reset_commands
+          @build_settings.call(self)
+        end
+      end
+
       def define_gem_init_builder
         file "#{build_dir}/gem_init.c" => [build.mrbcfile, __FILE__] + [rbfiles].flatten do |t|
           mkdir_p build_dir
@@ -303,6 +314,13 @@ module MRuby
 
         self
       end
+
+      private def reset_commands
+        MRuby::Build::COMMANDS.each do |command|
+          instance_variable_set("@#{command}", @build.send(command).clone)
+        end
+        @linker.run_attrs.each(&:clear)
+      end
     end # Specification
 
     class Version
@@ -408,6 +426,11 @@ module MRuby
           self.each(&:setup)
           gemset = self.setup_dependencies(build).keys.sort
         end until gemset == gemset_prev
+      end
+
+      def setup_build
+        each(&:setup_build)
+        self
       end
 
       def setup_dependencies(build)

--- a/lib/mruby/gem.rb
+++ b/lib/mruby/gem.rb
@@ -401,7 +401,16 @@ module MRuby
         end
       end
 
-      def generate_gem_table build
+      def setup(build)
+        gemset = nil
+        begin
+          gemset_prev = gemset
+          self.each(&:setup)
+          gemset = self.setup_dependencies(build).keys.sort
+        end until gemset == gemset_prev
+      end
+
+      def setup_dependencies(build)
         gem_table = each_with_object({}) { |spec, h| h[spec.name] = spec }
 
         default_gems = {}
@@ -423,6 +432,12 @@ module MRuby
             default_gems[dep[:gem]] ||= default_gem_params(dep)
           end
         end
+
+        gem_table
+      end
+
+      def generate_gem_table(build)
+        gem_table = setup_dependencies(build)
 
         each do |g|
           g.dependencies.each do |dep|

--- a/mrbgems/mruby-compiler/mrbgem.rake
+++ b/mrbgems/mruby-compiler/mrbgem.rake
@@ -3,15 +3,17 @@ MRuby::Gem::Specification.new 'mruby-compiler' do |spec|
   spec.author  = 'mruby developers'
   spec.summary = 'mruby compiler library'
 
-  objs = %w[codegen y.tab].map do |name|
-    src = "#{dir}/core/#{name}.c"
-    if build.cxx_exception_enabled?
-      build.compile_as_cxx(src)
-    else
-      objfile(src.pathmap("#{build_dir}/core/%n"))
+  spec.build_settings do
+    objs = %w[codegen y.tab].map do |name|
+      src = "#{dir}/core/#{name}.c"
+      if build.cxx_exception_enabled?
+        build.compile_as_cxx(src)
+      else
+        objfile(src.pathmap("#{build_dir}/core/%n"))
+      end
     end
+    build.libmruby_core_objs << objs
   end
-  build.libmruby_core_objs << objs
 end
 
 dir = __dir__

--- a/tasks/mrbgems.rake
+++ b/tasks/mrbgems.rake
@@ -2,8 +2,6 @@ MRuby.each_target do
   active_gems_txt = "#{build_dir}/mrbgems/active_gems.txt"
 
   if enable_gems?
-    # set up all gems
-    gems.each(&:setup)
     gems.check self
 
     # loader all gems

--- a/tasks/mrbgems.rake
+++ b/tasks/mrbgems.rake
@@ -2,6 +2,7 @@ MRuby.each_target do
   active_gems_txt = "#{build_dir}/mrbgems/active_gems.txt"
 
   if enable_gems?
+    gems.setup_build
     gems.check self
 
     # loader all gems

